shared/src/commonMain/kotlin/io/customer/shared/tracking/queue/QueueRunner.kt

```diff
@@ -0,0 +1,105 @@
+package io.customer.shared.tracking.queue
+
+import io.customer.shared.database.TrackingTask
+import io.customer.shared.database.TrackingTaskQueryHelper
+import io.customer.shared.sdk.meta.Workspace
+import io.customer.shared.tracking.api.*
+import io.customer.shared.tracking.api.model.*
+import io.customer.shared.tracking.constant.QueueTaskStatus
+import io.customer.shared.tracking.model.Task
+import io.customer.shared.tracking.model.TaskResponse
+import io.customer.shared.util.JsonAdapter
+import io.customer.shared.util.Logger
+import io.customer.shared.util.fromJSON
+
+/**
+ * Class responsible for sending tasks to server by communicating with network layer and updating
+ * tasks status and responses in database.
+ */
+internal interface QueueRunner {
+    suspend fun runQueueForTasks(pendingTasks: List<TrackingTask>): Result<Boolean>
+}
+
+internal class QueueRunnerImpl(
+    private val logger: Logger,
+    private val jsonAdapter: JsonAdapter,
+    private val workspace: Workspace,
+    private val trackingTaskQueryHelper: TrackingTaskQueryHelper,
+    private val trackingHttpClient: TrackingHttpClient,
+) : QueueRunner {
+    override suspend fun runQueueForTasks(
+        pendingTasks: List<TrackingTask>,
+    ): Result<Boolean> = kotlin.runCatching {
+        trackingTaskQueryHelper.updateTasksStatus(
+            status = QueueTaskStatus.SENDING,
+            tasks = pendingTasks,
+        )
+
+        val batchResult = processBatchTasks(pendingTasks = pendingTasks)
+        val batchResponse = batchResult.getOrNull()
+        if (batchResponse != null) {
+            val (response, tasksResponse) = batchResponse
+            trackingTaskQueryHelper.updateTasksStatusFromResponse(responses = tasksResponse)
+            return@runCatching !response.isServerUnavailable
+        } else {
+            logger.error("Failed to run queue for ${pendingTasks.size} tasks with error: ${batchResult.exceptionOrNull()?.message}")
+            trackingTaskQueryHelper.updateTasksStatus(
+                status = QueueTaskStatus.FAILED,
+                tasks = pendingTasks,
+            )
+            return@runCatching false
+        }
+    }
+
+    private suspend fun processBatchTasks(
+        pendingTasks: List<TrackingTask>,
+    ): Result<Pair<BatchTrackingResponse, List<TaskResponse>>> = kotlin.runCatching {
+        val result = trackingHttpClient.track(
+            tasks = pendingTasks.map { task ->
+                Task(
+                    identityType = workspace.identityType,
+                    profileIdentifier = task.identity,
```

Comment:
I was going through the PR and realized that [we are ignoring tasks which do not have identifier set](https://github.com/customerio/mobile-shared/blob/develop/shared/src/commonMain/sqldelight/io/customer/shared/database/TrackingTask.sq#L55-L59). The implementation is different than what we discussed and [suggested in initial proposal](https://www.notion.so/custio/Background-Queue-Implementation-3896eadbc7e34e0384a5cbec06b4fe84#407eb46548524c8ca34d5007942dc43c). 

Since we have retrying mechanism already, I would suggest not to exclude the tasks with empty/null identifier and let our retrying mechanism handle this by itself.

Reply:
Thanks for pointing this out Shahroz. I missed that completely 🤦🏻 I'll update it soon so the code matches the proposal.

Reply:
Adding note, we'll make these changes with anonymous tracking so that it easier to manage the changes.